### ISSUE TITLE
 Basic language extension as a built-in extension in VS Code

### ISSUE THREAD

**Issue by tamuratak** [James-Yu/LaTeX-Workshop#2731]
https://github.com/James-Yu/LaTeX-Workshop/commit/8dfaeb2cc314872293546fa453de352940b88bea#commitcomment-51335775

> I wanted to bring to your attention that setting "virtualWorkspaces": false will disable the whole extension, inlcuding grammars and language configuration contributions.
> To avoid this, we recommend to split off a basic language extension (grammars, language configuration, snippets) form the rich language support and have two extensions. Given the popularity of LateX, we would be willing to host such a basic language extension as a built-in extension in VS Code. Let me know if you are interested.
> See also https://github.com/microsoft/vscode/wiki/Virtual-Workspaces#disabling-a-language-extension

**Comment by aeschli** [comment]
Thanks for the issue. That's better.
We were discussing about where the grammars would be hosted, You currently have them in the same repo as the extension.
That's possible, but might be confusing as the extension, at some point, won't contain them any more (but rely on the built-in ones.

Do you prefer hosting the grammars where you have them now, or do you want to look into extracting them in a separate repo? The separate repo has the advantage that other editors/ides might want to consume that repo as well and you can easier invite other contributors to share the maintenance.

**Comment by jlelong** [comment]
Sorry for my silence, I am quite overworked these days and will not really be available before next week.

In the long term, I guess it would be clearer to host the grammars into a separate repo. 

I have a few questions

- Just to be sure, we would need to extract https://github.com/James-Yu/LaTeX-Workshop/tree/master/syntax for sure but is there anything else that should be moved too? I am thinking of https://github.com/James-Yu/LaTeX-Workshop/tree/master/snippets
- The language `id`'s are defined by the basic language extension, aren't they?
-  it is not clear how to handle the transition: new versions of the extensions (without the grammars) will not work with older VSCode versions with no LaTeX built-in grammars. We should probably keep a copy of the grammars in the extension repo for a while, shouldn't we?

**Comment by aeschli** [comment]
No worries, there's no rush. We are now in May endgame, so I won't have time myself in the next 1.5 weeks anyways.

The basic language extension would not contain any snippets. It's better they come from extensions. Users often disagree what a good snippet is and we don't want to be the judge on that. Also we don't have a mechanism to disable specific snippets that code from built-in extensions.
So I suggest to keep the snippets in your extension.

The language ids with file associations all go in the basic language extension

For backward compatibility, your extension can continue to define language id, configurations and grammar and your settings will win over the ones from the  basic language extension. Still the goal should be that after some time you can remove the duplicate contributions.

**Comment by tamuratak** [comment]
I have two concerns.

When the new extension becomes a built-in extension in VS Code, the number of users will increase explosively.  Especially, markdown users will use it with a LaTeX document embed in markdown  documents. They will request us tweaks for such usages. We will need to support them. It is a job I definitely want not to do. 

@jlelong, you are the one who actually does the job. If you are fine, it's OK.

The second concern is that we will receive lots of requests from users to change scope names. See #2126 and #2192.

**Comment by jlelong** [comment]
Grammars are very delicate objects mainly because of edge cases, which makes tweaking them very time consuming. I guess most users will be reasonable and I should be able to handle and discuss new requests. I have already spent a lot of time improving the original https://github.com/textmate/latex.tmbundle/blob/master/Syntaxes/LaTeX.plist and I am afraid that most requests will simply be out of reach.

Changing scope names is an other story as it induces visual changes and making users change their habits is difficult. It is clear that even if the grammars are built into VS Code, we will mainly be concerned by LW users experience.

All that being said, I still believe that having LaTeX grammar as a built-in extension in VS Code is a great opportunity. In the end, we develop and maintain LW on our free time and most people are grateful;

**Comment by coxackie** [comment]
Just discovered this discussion and I think it is very exciting. I think moving LaTeX grammar to a built-in extension in VS Code will help the general ecosystem, as other LaTeX-related extensions may also benefit. Hope this will be implemented soon!

**Comment by tamuratak** [comment]
@jlelong I think the following might be simple:

1. Create a subdirectory `./packages/latex-basics/` in this repository.
2. `git mv` syntax files of LaTeX Workshop to the directory. 
2. In the directory, you create a new VS Code extension only with syntax files.
3. We write the paths of the files in `package.json` of LaTeX Workshop and distribute LaTeX Workshop with the syntax files.

With this approach:

1. We can have the comprehensive git history for the new extension.
2. It might be easy to sync the syntax files between the new extension and LaTeX Workshop

**Comment by jlelong** [comment]
Actually, I have already started a new repository to hold the syntax files https://github.com/jlelong/vscode-latex-basics

I have relied on `git subtree` to keep the comprehensive git history of the syntax files. I think using a separate repo will make things easier once vscode provides `vscode-latex-basics` as a built-in extension. I suggest to keep a copy of the syntax files in LW until they are integrated into vscode. I still have a few things to fix but everything should be working very soon.

@James-Yu @tamuratak You should have received an invitation to join https://github.com/jlelong/vscode-latex-basics

**Comment by tamuratak** [comment]
I see.

**Comment by jlelong** [comment]
Here is the new repo holding the language ids, configurations and grammars https://github.com/jlelong/vscode-latex-basics
@aeschli @isidorn would you mind having a look at the new repo? Let me know if everything is fine (or not 🤔) for you to integrate it as a built-in extension.

**Comment by tamuratak** [comment]
I find a minor issue. See jlelong/vscode-latex-basics/issues/1

**Comment by tamuratak** [comment]
@jlelong  I have a question on `git subtree`.

Suppose that we want to remove tweaks of `autoClosingPairs` on `vscode-latex-basics`' side.

https://github.com/jlelong/vscode-latex-basics/blob/898c5155fc8a657df293ec6abc19890e9c0b3247/latex-language-configuration.json#L22-L48

We have to de-sync `latex-language-configuration.json` on both sides while syncing other files. Can we achieve that with `git subtree`?

**Comment by jlelong** [comment]
I have just used `git subtree` to extract a directory along with its comprehensive history to a new repo as described at [stackoverflow](https://stackoverflow.com/questions/359424/detach-move-subdirectory-into-separate-git-repository/17864475#17864475).

The syntax files in LW and `vscode-latex-basics` are not linked and cannot be easily linked using `git subtree` as the directory structure in `vscode-latex-basics` has been modified compared to the original LW. We must not modify the syntax and language configuration files in LW anymore. I will add a script to LW to automatically pull the syntaxes, language configurations and ids from `vscode-latex-basics` to keep everything synced. All this will eventually not be needed when he syntaxes, language configurations and ids are removed from LW.

**Comment by tamuratak** [comment]
> I will add a script to LW 

Thank you, I see.

> the syntaxes, language configurations and ids are removed from LW.

To my understanding, the following part exists to work with completions from LaTeX Workshop. Users of VS Code might not want it in `vscode-latex-basics`, and we might decide to delete it in the future. 

So, I think, at least, we have to keep it in LaTeX Workshop. 

```
	"autoClosingPairs": [
		["\\begin{", ""],
		["\\left(", ""],
		["\\left[", ""],
		["\\left{", ""],
		["\\bigl(", ""],
		["\\bigl[", ""],
		["\\bigl{", ""],
		["\\Bigl(", ""],
		["\\Bigl[", ""],
		["\\Bigl{", ""],
		["\\biggl(", ""],
		["\\biggl[", ""],
		["\\biggl{", ""],
		["\\Biggl(", ""],
		["\\Biggl[", ""],
		["\\Biggl{", ""],
		["\\(", ""],
		["\\[", ""],
		["\\{", ""],
		["@(", ""],
		["@[", ""],
		["@{", ""],
```

https://github.com/jlelong/vscode-latex-basics/blob/898c5155fc8a657df293ec6abc19890e9c0b3247/latex-language-configuration.json#L22-L48

**Comment by aeschli** [comment]
@jlelong Thanks a lot for separating out the new repo/extension! 

There are now two ways to proceed. 
a. You publish the extension as a basic language extension. it will appear on the marketplace for users to install. It's not a built-in extension.
b. We add a built-in basic latex extension to vscode with the same content as what you have now: grammars, language configuration, language definition. The only difference is that we have an update script that imports the grammars from your repo. You can change your repo to a grammar only repo: Change the name to something like `latext-tmlanguage`, remove the language configuration, remove the vscode specific stuff in package.json. You can check ou the [typescript grammars](https://github.com/Microsoft/TypeScript-TmLanguage) as an example

In a. you are in full control of the basic extension, but it's not built-in. Issues are filed against your extension repo.
In b. we own the basic extension, but rely on your input and update on your request and run the grammar update scripts on a regular basis. Issues are filed against vscode and we forward them to the grammar repo.

Let me know what you had in mind.

@alexr00 FIY

**Comment by jlelong** [comment]
@aeschli @alexr00 I had in mind approach b., but I have a few questions before actually making a choice

From our previous discussions on approach b., I thought you would also get the language ids and configurations from the new repo but it does not seem to be the case. How do you plan to get these files into VS Code? And more importantly, how could we ask for an update of them? These files are a little tricky and from time to time we have to improve them, see for instance #2985

In approach a.,
- Is it possible to require the installation of `vscode-latex-basics` from LaTeX-Workshop as `vscode-latex-basics` would be a requirement. 
- Could we overwrite the language configuration files of `vscode-latex-basics` with those of LaTeX-Workshop, see the [above comment](https://github.com/James-Yu/LaTeX-Workshop/issues/2731#issuecomment-981348890)? How to make sure it does not go the other way round? More generally, what is the general pattern to overwrite a language configuration?

**Comment by jlelong** [comment]
@tamuratak, I agree that https://github.com/jlelong/vscode-latex-basics/blob/898c5155fc8a657df293ec6abc19890e9c0b3247/latex-language-configuration.json#L22-L48 is related to our intellisense mechanism. 

Now, that all `@` snippets are actually provided as intellisense, I think we could get rid of this trick and just fix everything in our completion provider based on the value of `editor.autoClosingBrackets`. I shall give it a try soon.

**Comment by aeschli** [comment]
@jlelong For a), yes you would add `"extensionDependencies": ["James-Yu.vscode-latex-basics"] to the LaTeX-Workshop extension, This will make sure that vscode-latex-basics is automatically installed when LaTeX-Workshop is installed

I know that overwriting definitely works for b.). Installed extensions always beat the configurations from built-in extensions. For a.) I think it will be merged. Reading the code I'm not sure if these are ordered by dependency. Can you maybe try this out?

**Comment by jlelong** [comment]
> Reading the code I'm not sure if these are ordered by dependency

@aeschli Well this means that we cannot reliably overwrite the language definition provided by one extension in an other one. If the overwriting behaviour is not part of the specs, it could change anytime in the future. 

From our previous discussions on approach b., I thought you would also get the language ids and configurations from the new repo but it does not seem to be the case. 
- How do you plan to get these files into VS Code? 
- how could we update them? These files are a little tricky and from time to time we have to improve them, see for instance #2985

Although I would really like to see a built-in basic latex extension in vscode, the above questions are really important to decide which way to go.

**Comment by aeschli** [comment]
@jlelong We would start with the ones you have now. In case of an update, you can make a PR that we are happy to merge.

**Comment by jlelong** [comment]
Thanks for making all this clearer. I think the best approach is to have a built-in basic latex extension in vscode. With this approach,
- vscode will natively recognise the LaTeX language, which was the original motivation behind splitting LW
- the language configuration can be easily overwritten by LW or any other latex related extensions, whereas there is no reliable way of overwriting a language configuration provided by an other extension. This is of prime importance as discussed in https://github.com/James-Yu/LaTeX-Workshop/pull/2993#issuecomment-982317529

@tamuratak Is it fine for you? If so, I will continue talking with VS Code's PMs to figure out
- what parts of the language configuration files should be left in the basic latex extension
- how to move forward on microsoft/vscode-markdown-tm-grammar#99

**Comment by tamuratak** [comment]
I think it is fine.

**Comment by jlelong** [comment]
@aeschli Let us go for a built-in basic latex extension.

> You can change your repo to a grammar only repo: Change the name to something like `latex-tmLanguage`, remove the language configuration, remove the vscode specific stuff in package.json.

I have nothing against renaming the repo. Yet, I have the feeling that it would be simpler for us (especially for debugging purposes) to also keep the language configuration files and the `languages` and `grammars` sections of `package.json` in the repo even though you only automatically update the grammar files. I would like to start with keeping these files in the repo to make sure we have reached an acceptable reorganisation for LW. Would that be an issue for you? For instance, https://github.com/jeff-hykin/better-cpp-syntax, which is used to provide C/C++ grammars, contains the complete `package.json` file.

**Comment by aeschli** [comment]
Sure, no problem. We could also look into extending out grammar update script to include the language configuration.

**Comment by jlelong** [comment]
@aeschli You may also want to handle my PR microsoft/vscode-markdown-tm-grammar#99

In short, latex fenced code blocks in markdown are currently supported through a grammar injection from the latex grammar. When the latex grammar is part of vscode, it makes sense to treat latex as any other language and to handle latex fenced code blocks directly in the markdown grammar.

**Comment by aeschli** [comment]
Thanks, so we will add a new built-in extension. I will look into this next milestone, or if you want, make a PR.

**Comment by jlelong** [comment]
I will try to make a PR because we have to discuss/decide what should be going into the language configuration files of the built-in extension and what should be kept into LW.

**Comment by jlelong** [comment]
@aeschli @alexr00 I have made a PR attempt microsoft/vscode#138395.

**Comment by jlelong** [comment]
Closed by microsoft/vscode#138395.
